engine/userguide/storagedriver/btrfs-driver.md

```diff
@@ -227,179 +106,193 @@ loaded. To verify this, use the following command:
-recommended you perform your own tests before using this option.
-
-- **Use Data Volumes**. Data volumes provide the best and most predictable
-performance. This is because they bypass the storage driver and do not incur
-any of the potential overheads introduced by thin provisioning and
-copy-on-write. For this reason, you should place heavy write workloads on data
-volumes.
-
-- **Balance BTRFS**. Enable a cronjob to rebalance your BTRFS devices. e.g.
-Spread the subvolume's blocks evenly across your raid devices, and reclaim
-unused blocks. Without doing this, snapshots and subvolumes that docker
-removes will leave allocated blocks fillingup the BTRFS root volume. Once full
-you won't be able to re-balance, resulting in a potentially unrecoverable
-state without adding an additional storage device. If you would rather not
-automate this with crond, another option is to run a re-balance manually
-outside peak use times since the operation can be disk I/O intensive. This
-command will claim all chunks that are 1% used or less:
-
-  $ sudo btrfs filesystem balance start -dusage=1 /var/lib/docker
-
-  Dumping filters: flags 0x1, state 0x0, force is off
-  DATA (flags 0x2): balancing, usage=1
-  Done, had to relocate 673 out of 842 chunks
-
-More information on this topic can be read on the [BTRFS Wiki](https://btrfs.wiki.kernel.org/index.php/Balance_Filters#Balancing_to_fix_filesystem_full_errors).
+  filesystems like Btrfs. Many small random writes can compound this issue.
+  Fragmentation can manifest as CPU spikes when using SSDs or head thrashing
+  when using spinning disks. Either of these issues can harm performance.
+
+  If your Linux kernel version is 3.9 or higher, you can enable the `autodefrag`
+  feature when mounting a Btrfs volume. Test this feature on your own workloads
+  before deploying it into production, as some tests have shown a negative
+  impact on performance.
+
+- **SSD performance**: Btrfs includes native optimizations for SSD media.
+  To enable these features, mount the Btrfs filesystem with the `-o ssd` mount
+  option. These optimizations include enhanced SSD write performance by avoiding
+  optimization such as *seek optimizations* which do not apply to solid-state
+  media.
+
+- **Balance Btrfs filesystems often**: Use operating system utilities such as a
+  `cron` job to balance the Btrfs filesystem regularly, during non-peak hours.
+  This reclaims unallocated blocks and helps to prevent the filesystem from
+  filling up unnecessarily. You cannot rebalance a totally full Btrfs
+  filesystem unless you add additional physical block devices to the filesystem.
+  See the
+  [BTRFS Wiki](https://btrfs.wiki.kernel.org/index.php/Balance_Filters#Balancing_to_fix_filesystem_full_errors).
+
+- **Use fast storage**: Solid-state drives (SSDs) provide faster reads and
+  writes than spinning disks.
+
+- **Use volumes for write-heavy workloads**: Volumes provide the best and most
```

Comment:
Thinking of which; given that for BTRFS, the whole `/var/lib/docker` directory is BTRFS, volumes (which are stored in `/var/lib/docker/volumes` will still suffer from the performance penalties of BTRFS.

I know something has come up a few times, alternatives are;

- if following these instructions, use `/var/lib/docker/btrfs` as the location to mount the btrfs device (does that work?)
- or we should have an option to specify the location where volumes are stored